File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  // let set = new Set();
  //  set.add(arr[0]);
  // for i = 1 to arr.length - 1
  //  if set.has(arr[i]) return arr[i]
  // set.add(arr[i]);
  // end
  // return -1

  let set = new Set();
  set.add(arr[0]);

  for(let i = 1; i < arr.length; i++) {
    if(set.has(arr[i])) {
      return arr[i];
    } 
    set.add(arr[i]);
  }

  return -1;

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
def find_first_duplicate(arr)
  # type your code in here
  # const set = new Set()
  # arr.each do |item|
  #   if set.has(item) return item
  #   set.add(item)
  # end
  # return -1

  set = Set.new;
  
  arr.each do |item|
    return item if set.include?(item) 
    set << item
  end
  return -1

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
